src/core/store.ts

```diff
@@ -0,0 +1,653 @@
+import type { Atom, WritableAtom } from './atom'
+
+type AnyAtom = Atom<unknown>
+type AnyWritableAtom = WritableAtom<unknown, unknown>
+type OnUnmount = () => void
+type NonPromise<T> = T extends Promise<infer V> ? V : T
+type WriteGetter = Parameters<WritableAtom<unknown, unknown>['write']>[0]
+type Setter = Parameters<WritableAtom<unknown, unknown>['write']>[1]
+
+const hasInitialValue = <T extends Atom<unknown>>(
+  atom: T
+): atom is T & (T extends Atom<infer Value> ? { init: Value } : never) =>
+  'init' in atom
+
+const IS_EQUAL_PROMISE = Symbol()
+const INTERRUPT_PROMISE = Symbol()
+type InterruptablePromise = Promise<void> & {
+  [IS_EQUAL_PROMISE]: (p: Promise<void>) => boolean
+  [INTERRUPT_PROMISE]: () => void
+}
+
+const isInterruptablePromise = (
+  promise: Promise<void>
+): promise is InterruptablePromise =>
+  !!(promise as InterruptablePromise)[INTERRUPT_PROMISE]
+
+const createInterruptablePromise = (
+  promise: Promise<void>
+): InterruptablePromise => {
+  let interrupt: (() => void) | undefined
+  const interruptablePromise = new Promise<void>((resolve, reject) => {
+    interrupt = resolve
+    promise.then(resolve, reject)
+  }) as InterruptablePromise
+  interruptablePromise[IS_EQUAL_PROMISE] = (p: Promise<void>) =>
+    p === interruptablePromise || p === promise
+  interruptablePromise[INTERRUPT_PROMISE] = interrupt as () => void
+  return interruptablePromise
+}
+
+type Revision = number
+type InvalidatedRevision = number
+type ReadDependencies = Map<AnyAtom, Revision>
+
+// immutable atom state
+export type AtomState<Value = unknown> = {
+  e?: Error // read error
+  p?: InterruptablePromise // read promise
+  c?: () => void // cancel read promise
+  w?: Promise<void> // write promise
+  v?: NonPromise<Value>
+  r: Revision
+  i?: InvalidatedRevision
+  d: ReadDependencies
+}
+
+type Listeners = Set<() => void>
+type Dependents = Set<AnyAtom>
+type Mounted = {
+  l: Listeners
+  d: Dependents
+  u: OnUnmount | void
+}
+
+// for debugging purpose only
+type StateListener = (updatedAtom: AnyAtom, isNewAtom: boolean) => void
+
+// store methods
+export const GET_VERSION = 'v'
```

Comment:
What about using Symbols?

Reply:
Symbols are a bit worse for minification. Besides, it's not very TS friendly.